File: services/service.py
from typing import List
from repositories import (
    CatRepository,
    DogRepository,
)
from repositories.repository import DEFAULT_LIST_LIMIT
from models import (
    CatResponse,
    Cat,
    DogResponse,
    Dog,
)


class CatService:
    def __init__(self, repository: CatRepository):
        self.repository = repository

    async def get_by_id(self, item_id: str) -> CatResponse:
        return await self.repository.get_by_id(item_id)

    async def get_list(self, limit: int = DEFAULT_LIST_LIMIT) -> List[CatResponse]:
        return await self.repository.get_list(limit)

    async def create(self, item: Cat) -> CatResponse:
        return await self.repository.create(item)

    async def update(self, item: Cat) -> CatResponse:
        return await self.repository.update(item)

    async def soft_delete(self, item_id: str):
        await self.repository.delete(item_id)


class DogService:
    def __init__(self, repository: DogRepository):
        self.repository = repository

    async def get_by_id(self, item_id: str) -> DogResponse:
        return await self.repository.get_by_id(item_id)

    async def get_list(self, limit: int = DEFAULT_LIST_LIMIT) -> List[DogResponse]:
        return await self.repository.get_list(limit)

    async def create(self, item: Dog) -> DogResponse:
        return await self.repository.create(item)

    async def replace(self, item: Dog) -> DogResponse:
        return await self.repository.replace(item)

    async def soft_delete(self, item_id: str):
        await self.repository.delete(item_id)

File: services/service.ts
import {
    CatRepository,
    DogRepository,
} from '@repositories';
import {
    Cat,
    CatResponse,
    CatResponseSchema,
    CatEntitySchema,
    CatUpdate,
    Dog,
    DogResponse,
    DogResponseSchema,
    DogEntitySchema,
    DogUpdate,
} from '@models';

export class CatService {
  private _repo: CatRepository;

  constructor(repo: CatRepository) {
    this._repo = repo;
  }

  create = async (item: Cat): Promise<CatResponse> => {
    const newItem = CatEntitySchema.parse(item);
    const createdItem: CatResponse = await this._repo.create(newItem);
    return CatResponseSchema.parse(createdItem);
  };

  get = async (id: string): Promise<CatResponse> => {
    const item = await this._repo.get(id);
    return CatResponseSchema.parse(item);
  };

  list = async (limit?: number): Promise<CatResponse[]> => {
    const items = await this._repo.list(limit);
    return items.map((item) => CatResponseSchema.parse(item));
  };

  update = async (item: CatUpdate): Promise<CatResponse> => {
    const updatedItem = await this._repo.update({ ...item, updatedTimestamp: new Date().toISOString() });
    return CatResponseSchema.parse(updatedItem);
  };

  delete = async (id: string): Promise<void> => this._repo.delete(id);
}

export class DogService {
  private _repo: DogRepository;

  constructor(repo: DogRepository) {
    this._repo = repo;
  }

  create = async (item: Dog): Promise<DogResponse> => {
    const newItem = DogEntitySchema.parse(item);
    const createdItem: DogResponse = await this._repo.create(newItem);
    return DogResponseSchema.parse(createdItem);
  };

  get = async (id: string): Promise<DogResponse> => {
    const item = await this._repo.get(id);
    return DogResponseSchema.parse(item);
  };

  list = async (limit?: number): Promise<DogResponse[]> => {
    const items = await this._repo.list(limit);
    return items.map((item) => DogResponseSchema.parse(item));
  };

  replace = async (item: DogUpdate): Promise<DogResponse> => {
    const existing = await this._repo.get(item.id);
    const entity = DogEntitySchema.parse({
      ...item,
      id: existing.id,
      isDeleted: false,
      createdTimestamp: existing.createdTimestamp,
      createdBy: existing.createdBy,
      updatedTimestamp: new Date().toISOString(),
    });
    const replacedItem = await this._repo.replace(entity);
    return DogResponseSchema.parse(replacedItem);
  };

  delete = async (id: string): Promise<void> => this._repo.delete(id);
}
